lib/hotel_manager.rb

```diff
@@ -0,0 +1,103 @@
+require_relative "room.rb"
+require_relative "reservation.rb"
+require_relative "block.rb"
+
+module Hotel
+  class HotelManager
+    attr_reader :rooms, :reservations, :blocks
+
+    def initialize(rooms)
+      @rooms = rooms
+      @reservations = []
+      @blocks = []
+    end
+
+    #reservation methods
+    def make_reservation(room_number, start_date, end_date)
+      room = find_room_by_number(room_number)
+      if self.list_available_rooms(start_date, end_date).include?(room)
+        new_reservation = Hotel::Reservation.new(room: room, start_date: start_date, end_date: end_date)
+        @reservations.push(new_reservation)
+        return new_reservation
+      else
+        raise ArgumentError, "Room #{room_number} is unavailable. Must use an available room."
+      end
+    end
+
+    def list_reservations_by_date(date)
+      date = Date.parse(date)
+      found_reservations = []
+      @reservations.each do |reservation|
+        if date >= reservation.start_date && date < reservation.end_date
+          found_reservations.push(reservation)
+        end
+      end
+      return found_reservations
+    end
+
+    #block methods
+    def make_block(room_numbers, start_date, end_date, discount_rate)
+      start_date, end_date = parse_dates(start_date, end_date)
+      block_rooms = room_numbers.map { |number| find_room_by_number(number) }
+      validate_block(block_rooms, start_date, end_date)
+      new_block = Hotel::Block.new(rooms: block_rooms, start_date: start_date, end_date: end_date, discount_rate: discount_rate)
+      @blocks.push(new_block)
+      return new_block
+    end
+
+    def reserve_from_block(block, room_number)
+      room = find_room_by_number(room_number)
+      block.book_room(room)
+      new_reservation = Hotel::Reservation.new(room: room, start_date: block.start_date, end_date: block.end_date)
+      @reservations.push(new_reservation)
+      return new_reservation
+    end
+
+    #room methods
+    def list_available_rooms(start_date, end_date)
+      start_date, end_date = parse_dates(start_date, end_date)
+      available_rooms = @rooms
+      reservations.each do |reservation|
+        if reservation.start_date >= start_date && reservation.end_date <= end_date
+          available_rooms.delete(reservation.room)
```

Comment:
Here the `Hotel` goes in and looks at the details of the `Reservation` class (its start date and end date), which is not great design. `Hotel` needs to know a lot about `Reservation` for that to work, and the two classes are more tightly coupled than they need to be. An alternative approach would be to write a method `Reservation#overlaps?(start_date, end_date)`, and call that here instead:

```ruby
reservations.each do |reservation|
  if reservation.overlaps?(start_date, end_date)
    available_rooms.delete(reservation.room)
  end
end
```

I know we hadn't read POODR ch 4 yet when you wrote this code, but this is what Metz is talking about when she says you should "ask for _what_ instead of telling _how_".